lib/client_generator.dart

```diff
@@ -34,11 +34,15 @@ class ClientApiGenerator {
     var outputType = service._getDartClassName(m.outputType);
     out.addBlock(
         'Future<$outputType> $methodName('
-        'ClientContext ctx, $inputType request) {',
+        'ClientContext ctx, $inputType request) async {',
         '}', () {
       out.println('var emptyResponse = new $outputType();');
-      out.println('return _client.invoke(ctx, \'${service._descriptor.name}\', '
+      out.println(
+          'final response = await _client.invoke(ctx, \'${service._descriptor.name}\', '
           '\'${m.name}\', request, emptyResponse);');
+
+      out.println('emptyResponse.mergeFromMessage(response);');
```

Comment:
You don't need to merge `response` into `emptyResponse`. This is already done in `_client.invoke()`. You could simplify the generated code to something like:

```dart
final response = new $outputType();
await _client.invoke(ctx, ..., response);
return response;
```

Reply:
Thanks for your reply and nice comment.
What about I change the RpcClient 
FROM:
``` dart
abstract class RpcClient {
  /// Sends a request to a server and returns the reply.
  ///
  /// The implementation should serialize the request as binary or JSON, as
  /// appropriate. It should merge the reply into [emptyResponse] and
  /// return it.
  Future<GeneratedMessage> invoke(
      ClientContext ctx,
      String serviceName,
      String methodName,
      GeneratedMessage request,
      GeneratedMessage emptyResponse);
}
```
TO
``` dart
abstract class RpcClient {
  /// Sends a request to a server and returns the reply.
  ///
  /// The implementation should serialize the request as binary or JSON.
  Future<GeneratedMessage> invoke(
      ClientContext ctx,
      String serviceName,
      String methodName,
      GeneratedMessage request);
}
```

I think the user who implement RpcClient don't need to know about the mergeFromMessage method, just deserialize a new reponse and return it.

What about your opinion?

Reply:
The `RpcClient` interface is a lot harder to change, since there's lots of code depending on it (or implementing it). Plus, the way you deserialize a Protobuf message is by merging into an empty instance of the desired message type.

Passing in an empty response object is the way we tell `invoke()` how to deserialize the response. The other option would be to have `invoke()` return a `List<int>` that the generated client stubs would then have to deserialize (by merging), so we don't really win anything.

I think the simplest solution is to leave `RpcClient` as it is, and fix the generated code. If you're worried about not returning the result of `await _client.invoke()`, you could probably cast it:

```dart
final response = await _client.invoke(ctx, ..., new $outputType());
return response as $outputType;
```

But the documentation for `RpcClient.invoke` says that it should merge the response into the provided object and return it.

Reply:
Is there any other problem to merging this PR?